README.md
# ngx-fab

Simple floating action buttons for Angular >= 2

[Demo can be found here](https://hughjdavey.github.io/ngx-fab)

### Installation

* `npm install --save ngx-fab`
* Edit your app.module file:

``` typescript
...
import { NgxFabModule } from 'ngx-fab';

@NgModule({
  ...
  imports: [
    ...
    NgxFabModule
  ],
  ...
})
```

### Usage

##### Type Definitions:

* `type FabSize = 'small' | 'medium' | 'large'`
* `type FabPosition = 'top-left' | 'top-middle' | 'top-right' | 'middle-left' | 'middle-middle' | 'middle-right' | 'bottom-left' | 'bottom-middle' | 'bottom-right'`

##### `@Input()` options:

* `backgroundColor` [string] - HTML color name or hexcode for background color (defaults to 'crimson')
* `fabSize` [FabSize] - fab size, 40, 60 or 100px respectively (defaults to 'medium')
* `iconClass` [string] - font awesome icon classes from 'fa', 'fab', 'far' or 'fas' collections (defaults to 'fa fa-plus')
* `iconColor` [string] - HTML color name or hexcode for icon color (defaults to 'black' or 'white' depending on `backgroundColor`)
* `position` [FabPosition] - position of fab within container (defaults to 'bottom-right')
* `additionalX` [float] - additional positive or negative offset on the x-axis used for custom positioning (defaults to 0)
* `additionalY` [float] - additional positive or negative offset on the y-axis used for custom positioning (defaults to 0)

### Examples

##### deafault FAB - bottom right, crimson background, plus icon
* `<ngx-fab></ngx-fab>`

##### custom background color, icon and position
* `<ngx-fab [backgroundColor]="'#88B04B'" [iconClass]="'fas fa-atom'" [position]="'top-middle'"></ngx-fab>`

##### custom icon and icon color
* `<ngx-fab [backgroundColor]="'yellow'" [iconClass]="'fas fa-plus'" [iconColor]="'#ff00ff'"></ngx-fab>`

##### custom sizes
* `<ngx-fab [backgroundColor]="'crimson'" [iconClass]="'fas fa-star'" [fabSize]="'small'"></ngx-fab>`
* `<ngx-fab [backgroundColor]="'crimson'" [iconClass]="'fas fa-star'" [fabSize]="'large'"></ngx-fab>`

##### custom positioning via `additionalX` to form a row of 3
* `<ngx-fab [iconClass]="'fas fa-star'" [position]="'middle-middle'" [fabSize]="'small'" [additionalX]="-50"></ngx-fab>`
* `<ngx-fab [iconClass]="'fas fa-star'" [position]="'middle-middle'" [fabSize]="'small'"></ngx-fab>`
* `<ngx-fab [iconClass]="'fas fa-star'" [position]="'middle-middle'" [fabSize]="'small'" [additionalX]="50"></ngx-fab>`


CHANGELOG.md
# Changelog

## 1.0.0 (15/08/2018)

* Initial release


src/lib/ngx-fab.component.ts
import { Component, ElementRef, Input } from '@angular/core';

@Component({
  selector: 'ngx-fab',
  templateUrl: './ngx-fab.component.html',
  styleUrls: [ './ngx-fab.component.css' ],
})
export class NgxFabComponent {

  @Input()
  backgroundColor: string = 'crimson';

  @Input()
  fabSize: FabSize = 'medium';

  @Input()
  iconClass: string = 'fa fa-plus';

  @Input()
  iconColor: string;

  @Input()
  position: FabPosition = 'bottom-right';

  @Input()
  additionalX: number = 0;

  @Input()
  additionalY: number = 0;

  constructor(private elementRef: ElementRef) { }

  background(): Object {
    return {
      'background-color': this.backgroundColor,
      'border-color': this.backgroundColor,
    };
  }

  icon(): string {
    return `${this.iconClass} ${this.fabSize === 'large' ? 'fa-2x' : this.fabSize === 'medium' ? 'fa-lg' : ''}`;
  }

  color(): string {
    return this.iconColor || this.defaultIconColor();
  }

  size(): string {
    return `fab-${this.fabSize}`;
  }

  positionStyle(): Object {
    const parentBoundingRect = this.elementRef.nativeElement.parentElement.getBoundingClientRect();
    const fabWidth = this.fabSize === 'small' ? 40 : this.fabSize === 'medium' ? 60 : 100;
    const desiredOffsetX = (fabWidth / 3) + this.additionalX;
    const desiredOffsetY = (fabWidth / 3) + this.additionalY;

    const leftLeft = `${parentBoundingRect.left + desiredOffsetX}px`;
    const leftMid = `${parentBoundingRect.left + (parentBoundingRect.width / 2) - (fabWidth / 2) + this.additionalX}px`;
    const leftRight = `${parentBoundingRect.right - (fabWidth + desiredOffsetX)}px`;

    const topTop = `${parentBoundingRect.top + desiredOffsetY}px`;
    const topMid = `${parentBoundingRect.top + (parentBoundingRect.height / 2) - (fabWidth / 2) + this.additionalY}px`;
    const topBottom = `${parentBoundingRect.bottom - (fabWidth + desiredOffsetY)}px`;

    switch (this.position) {
      case 'top-left': return { top: topTop, left: leftLeft };
      case 'top-middle': return { top: topTop, left: leftMid };
      case 'top-right': return { top: topTop, left: leftRight };
      case 'middle-left': return { top: topMid, left: leftLeft };
      case 'middle-right': return { top: topMid, left: leftRight };
      case 'middle-middle': return { top: topMid, left: leftMid };
      case 'bottom-left': return { top: topBottom, left: leftLeft };
      case 'bottom-middle': return { top: topBottom, left: leftMid };
      case 'bottom-right': return { top: topBottom, left: leftRight };
    }
  }

  // inspired by this SO answer https://stackoverflow.com/a/3943023
  private defaultIconColor(): string {
    const colorHex = this.standardizeColor(this.backgroundColor);

    if (!colorHex || colorHex.length !== 7) {
      return 'white';
    }

    const sRGBs = [ colorHex.substring(1, 3), colorHex.substring(3, 5), colorHex.substring(5, 7) ].map(s => parseInt(s, 16));
    return (sRGBs[0] * 0.299 + sRGBs[1] * 0.587 + sRGBs[2] * 0.114) > 186 ? 'black' : 'white';
  }

  // inspired by JayB's comment on this SO answer https://stackoverflow.com/a/24390910
  private standardizeColor(colorName): string {
    const ctx = document.createElement('canvas').getContext('2d');
    ctx.fillStyle = colorName;
    return ctx.fillStyle.toString();
  }
}

export type FabPosition = 'top-left' | 'top-middle' | 'top-right' | 'middle-left' | 'middle-middle' | 'middle-right' | 'bottom-left' | 'bottom-middle' | 'bottom-right';

export type FabSize = 'small' | 'medium' | 'large';
